src/listpack.c

```diff
@@ -732,12 +732,6 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *ele, uint32_t size, un
                 old_listpack_bytes-poff-replaced_len);
     }
 
-    /* Realloc after: we need to free space. */
-    if (new_listpack_bytes < old_listpack_bytes) {
-        if ((lp = lp_realloc(lp,new_listpack_bytes)) == NULL) return NULL;
-        dst = lp + poff;
-    }
-
```

Comment:
i don't think we want to remove these lines.
keep in mind that listpack is a generic data stracture that can be some day used for many things (we intend to replace all our ziplists).
we need to make sure it use efficient in streams, but we don't want other users to suffer from remaining with huge overheads.
in some way we need to let the caller control when to over allocate, and when to trim or avoid trimming.

Reply:
Off topic: Although listpacks are better than ziplists, we can do even better by using the suffix bytes for payload and size bits, rather than only representing the total size. I have an encoding in mind where a 7bit integer is only 1 byte including suffix (0xxxxxxx), a 12bit integer is 2 bytes (10xxxxxx 10xxxxxx), 16bit integer is 3 bytes (1100xxxx xxxxxxxx 1100xxxx), short strings have only 2 bytes overhead with 4 bits for size in prefix and suffix (1101ssss <data> 1101ssss) and so on. I'll post it somewhere soon if you're interested.

Reply:
reverted ;D